Update sampling code, add exponential (biased) sampling

Bounciness opened the pull request:
This commit reorganized the sampling code into a submodule in external/Volume-and-Sampling/, which is maintained here: https://github.com/Bounciness/Volume-and-Sampling. Before this merge, I was separately maintaining two versions of the sampling code (one for COBRA, one for general use). This update merges the two functionalities. If the COBRA wrappers are used, everything should behave identically to before.

The main new feature is that exponential (biased) sampling of the solution space is supported. This update includes a few minor bug fixes with the sampling code. 

**I hereby confirm that I have:**

- [X] Tested my code on my own machine
- [X] Followed the guidelines in the [Contributing Guide](https://opencobra.github.io/cobratoolbox/docs/contributing.html)
- [X] Selected `develop` as a target branch (top left drop-down menu)

*(Note: You may replace [] with [X] to check the box)*

tpfau commented on src/analysis/sampling/CHRR/chrrSampler.m:
```diff
@@ -1,45 +1,40 @@
-% .. Authors: - Ben Cousins and Hulda S. Haraldsdóttir, March 2017
+% samples ... n x numSamples matrix of random flux samples
+% roundedPolytope ... The rounded polytope. Save for use in subsequent
+%                     rounds of sampling.
```
Please keep to the defined Documentation scheme detailed in [this](https://github.com/opencobra/cobratoolbox/blob/master/docs/source/guides/documentationGuide.md) guide

tpfau commented on src/analysis/sampling/CHRR/chrrSampler.m:
```diff
@@ -1,45 +1,40 @@
-function [samples, roundedPolytope, minFlux, maxFlux] = chrrSampler(model, numSkip, numSamples, toRound, roundedPolytope, minFlux, maxFlux, useFastFVA)
-% Generate uniform random flux samples with CHRR
-% Coordinate Hit-and-Run with Rounding
+function [samples,roundedPolytope] = chrrSampler(model,numSkip,numSamples,toRound,roundedPolytope,minFlux,maxFlux)
```
If it is necessary to change the header (which is strongly discouraged), please make sure, that any tutorials and tests are updated accordingly. 
From what I see in the code, this change essentially removed the information in `P.A_eq` and `P.b_eq`. Is that intended? As outputs might change by this alteration.

tpfau commented on src/analysis/sampling/CHRR/chrrParseModel.m:
```diff
@@ -1,34 +1,29 @@
+% October 2016, Ben Cousins and Hulda S. Haraldsdóttir
+%
+% --Moved objective function handling to preprocess function: Ben Cousins,
+% November 2017
```
Please format the documentation according to the [documentation guide](https://github.com/opencobra/cobratoolbox/blob/master/docs/source/guides/documentationGuide.md)

tpfau commented on src/analysis/sampling/CHRR/chrrExpSampler.m:
```diff
@@ -0,0 +1,125 @@
+% roundedPolytope ... The rounded polytope. Save for use in subsequent
+%                     rounds of sampling.
+%
+% November 2017, Ben Cousins and Hulda S. Haraldsdóttir
```
please format the documentation according to the [documentation guide](https://github.com/opencobra/cobratoolbox/blob/master/docs/source/guides/documentationGuide.md)

Bounciness replied:
This should be fixed now.

tpfau commented on src/analysis/sampling/CHRR/chrrSampler.m:
```diff
@@ -78,33 +69,30 @@
-    
-    P.A_eq = [P.A_eq; eq_constraints];
-    P.b_eq = [P.b_eq; minFlux(isEq)];
+%     if toGetWidths
```
Please remove large commented sections.

Bounciness replied:
Removed.

tpfau commented:
Oh, also, could you add a test for the exponential sampling method?

laurentheirendt commented:
@Bounciness ?

Bounciness commented:
Thanks a lot for your feedback Thomas, sorry for delay getting back to you. I've added a test for exponential sampling in analysis/testSampling/testSampleCbModel.m. 

With regards to the function header being different, the restructuring of the code makes the previous function header unnecessarily complex (i.e. we can get rid of two arguments). I ran the sampling tutorial line-by-line, everything seems to be good. The only real change required was changing the function call to chrrSampler() in sampling/sampleCbModel.m. I also added a function call to chrrExpSampler() in sampleCbModel.m. If a user is doing sampling through sampleCbModel(), then everything should be equivalent to before.

Hopefully all the documentation guidelines are adhered to, let me know if there's anything I missed.

tpfau commented:
Great, 
Would it be posible to add ethe exponential sampler as an option to sampleCbModel?
And could you update the tutorial?

rmtfleming commented:
@Gpreciat is responsible for maintaining the tutorial, which can be updated once we have fully tested this new sampler functionality.

Bounciness commented:
I've added CHRR_EXP as an option to sampleCbModel.

laurentheirendt commented:
@Bounciness, may you solve the conflict here? Thanks 👍

Bounciness commented:
@laurentheirendt I seem to be stuck in an infinite loop of conflicts. I thought I resolved it through the web interface on Github but it still says conflicting. Any suggestions?

laurentheirendt commented:
@Bounciness, I worked on this PR and I submitted [this PR](https://github.com/Bounciness/Volume-and-Sampling/pull/1) to make it work.

Once this is in, I will continue working on this PR. Thanks 👍

Bounciness commented:
@laurentheirendt, I'm terribly sorry! I forgot to commit my updates to preprocess.m on my submodule Volume-and-Sampling, which is why you were getting the errors. 

I've pushed my updated version to my submodule (which is essentially the version you submitted in your PR, but before calling fastFVA, it checks for CPLEX--this is the existing behavior in COBRA for sampling).

laurentheirendt commented:
@Bounciness, no worries. I will close this PR and resubmit the updated one. Thanks 👍
